pmlc/dialect/stripe/into_mlir.cc

```diff
@@ -426,9 +500,20 @@ static mlir::FuncOp ProgramIntoMLIR(MLIRContext* ctx, const stripe::Block& block
   for (const auto& ref : block.refs) {
     auto argIndex = argcnt++;
     auto arg = func.getArgument(argIndex);
-    Type tensorRefType = ShapeIntoTensorRefType(ctx, ref.interior_shape);
+    Type tensorRefType = ShapeIntoTensorRefType(ctx, ref.interior_shape, ref.location);
     auto tensorRefOp = builder.create<TensorRefOp>(loc, tensorRefType, arg);
-    initial.refs.emplace(ref.into(), tensorRefOp);
+    bool any_non_zero_offsets = false;
+    // N.B. initial.idxs is empty here, but that's reasonable in this case; these incoming parameter
+    // refinement locations shouldn't be dependent on any external indicies.
+    auto offsets = LocationIntoTensorOffsets(&builder, initial.idxs, ref.location, &any_non_zero_offsets);
+    if (any_non_zero_offsets) {
+      auto zero = builder.create<AffineConstOp>(loc, builder.getType<AffineType>(), builder.getI64IntegerAttr(0));
+      offsets.resize(offsets.size() + ref.interior_shape.dims.size(), zero);
+      auto refOp = builder.create<RefineOp>(loc, tensorRefOp.getType(), tensorRefOp, offsets);
+      initial.refs.emplace(ref.into(), refOp);
+    } else {
+      initial.refs.emplace(ref.into(), tensorRefOp);
+    }
```

Comment:
Can we have `LocationIntoTensorOffsets` actually be `LocationIntoTensorOps` so that it can return a `tensor_ref` or a `refine` depending on any non zero offsets. I'm worried about splitting the logic like this.

Reply:
That turns out to be tricky, since two of the places where it's used also need to add in the non-location offsets when they're creating the `RefineOp`.